mm2src/coins/qrc20/swap_ops.rs

```diff
@@ -0,0 +1,917 @@
+                Some(token) => return ERR!("Payment tx 'token_address' arg is invalid, found {:?}", token),
+                None => return ERR!("Couldn't find 'token_address' in erc20Payment call"),
+            };
+
+            let receiver = match decoded.next() {
+                Some(Token::Address(addr)) => addr,
+                Some(token) => return ERR!("Payment tx 'receiver' arg is invalid, found {:?}", token),
+                None => return ERR!("Couldn't find 'receiver' in erc20Payment call"),
+            };
+
+            let secret_hash = match decoded.next() {
+                Some(Token::FixedBytes(hash)) => hash,
+                Some(token) => return ERR!("Payment tx 'secret_hash' arg is invalid, found {:?}", token),
+                None => return ERR!("Couldn't find 'secret_hash' in erc20Payment call"),
+            };
+
+            let timelock = match decoded.next() {
+                Some(Token::Uint(t)) => t,
+                Some(token) => return ERR!("Payment tx 'timelock' arg is invalid, found {:?}", token),
+                None => return ERR!("Couldn't find 'timelock' in erc20Payment call"),
+            };
+
+            let mut events = try_s!(transfer_events_from_receipt(&receipt)).into_iter();
+            let event = match events.next() {
+                Some(e) => e,
+                None => return ERR!("Couldn't find 'Transfer' event from logs"),
+            };
+            // check if the erc20Payment emitted only one Transfer event
+            if events.next().is_some() {
+                return ERR!("'erc20Payment' should emit only one 'Transfer' event");
+            }
+
+            if event.contract_address != self.contract_address {
+                return ERR!(
+                    "Unexpected token address {:#02x} in 'Transfer' event, expected {:#02x}",
+                    event.contract_address,
+                    self.contract_address
+                );
+            }
+            if event.amount != value {
+                return ERR!(
+                    "Unexpected amount {} in 'Transfer' event, expected {}",
+                    event.amount,
+                    value
+                );
+            }
+
+            let contract_address_from_script = try_s!(extract_token_addr_from_script(&script_pubkey));
+            // `erc20Payment` function should emit a `Transfer` event where the receiver is the swap contract
+            if event.receiver != contract_address_from_script {
+                return ERR!(
+                    "Contract address {:#02x} from script pubkey and receiver {:#02x} in 'Transfer' event are different",
+                    contract_address_from_script,
+                    event.receiver
+                );
+            }
+
+            return Ok(Erc20PaymentDetails {
+                output_index: receipt.output_index,
+                swap_id,
+                value,
+                token_address,
+                swap_contract_address: event.receiver,
```

Comment:
Even despite `event.receiver` is compared to `contract_address_from_script` let's use `contract_address_from_script` explicitly here to not create an impression that it might be correct to use `event.receiver` to derive the `swap_contract_address`.

Reply:
Replaced `swap_contract_address: event.receiver,` with `swap_contract_address: contract_address_from_script,`